pkg/bootstrap/node.go

```diff
@@ -745,57 +753,179 @@ func checkCoreosVersion() error {
 	return nil
 }
 
-func pullRawCoreosImage() error {
-	var cmdPath string
-	var err error
+func ensureCoreosVersion() {
+	if err := checkCoreosVersion(); err != nil {
+		log.Println(err)
+		log.Fatalf("You will need to remove the image by 'sudo machinectl remove coreos' then the next run of kube-spawn will download version %s of coreos image automatically.", coreosStableVersion)
+	}
+}
 
-	// TODO: use machinetool pkg
-	if cmdPath, err = exec.LookPath("machinectl"); err != nil {
-		// fall back to an ordinary abspath to machinectl
-		cmdPath = "/usr/bin/machinectl"
+func PrepareCoreosImage(ImageGpgVerify bool) error {
+	// If no coreos image exists, just download it
+	if !machinetool.ImageExists("coreos") {
+		log.Printf("pulling coreos image...")
+		if err := pullRawCoreosImage(ImageGpgVerify); err != nil {
+			return err
+		}
+	} else {
+		// If coreos image is not new enough, remove the existing image,
+		// then next time `kube-spawn up` will download a new image again.
+		ensureCoreosVersion()
 	}
+	return nil
+}
 
-	args := []string{
-		cmdPath,
-		"pull-raw",
-		"--verify=no",
-		"https://alpha.release.core-os.net/amd64-usr/current/coreos_developer_container.bin.bz2",
-		"coreos",
+func downloadFile(dest, url string) error {
+	f, err := os.Create(dest)
+	if err != nil {
+		return err
 	}
+	defer f.Close()
 
-	cmd := exec.Cmd{
-		Path:   cmdPath,
-		Args:   args,
+	response, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	progress := &ioprogress.Reader{
+		Reader: response.Body,
+		Size:   response.ContentLength,
+	}
+
+	if _, err := io.Copy(f, progress); err != nil {
+		return err
+	}
+
+	if err := f.Sync(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func downloadImage() error {
+	return downloadFile(imageTmpFile, imageUrl)
+}
+
+func downloadSignature() error {
+	return downloadFile(signatureTmpFile, signatureUrl)
+}
+
+func verifyImage() error {
+	// TODO(nhlfr): Try to use or implement some library for managing PGP keys instead
+	// of executing gpg binary.
+	gpgCmdPath, err := exec.LookPath("gpg")
+	if err != nil {
+		gpgCmdPath = "/usr/bin/gpg"
```

Comment:
Some systems have both `gpg` and `gpg2`. Not sure picking `/usr/bin/gpg` is always right.